pygsti/tools/matrixtools.py

```diff
@@ -369,27 +413,28 @@ def independent_columns(m, initial_independent_cols=None, tol=1e-7):
     list
         A list of the independent-column indices of `m`.
     """
-    indep_cols = []
-
     if not _sps.issparse(m):
-        running_indep_cols = initial_independent_cols.copy() \
-            if (initial_independent_cols is not None) else _np.empty((m.shape[0], 0), m.dtype)
-        num_indep_cols = running_indep_cols.shape[0]
-
-        for j in range(m.shape[1]):
-            trial = _np.concatenate((running_indep_cols, m[:, j]), axis=1)
-            if _np.linalg.matrix_rank(trial, tol=tol) == num_indep_cols + 1:
-                running_indep_cols = trial
-                indep_cols.append(j)
-                num_indep_cols += 1
+        if initial_independent_cols is None:
+            proj_m = m.copy()
+        else:
+            assert initial_independent_cols.shape[0] == m.shape[0]
+            q = _spl.qr(initial_independent_cols, mode='econ')[0]
+            # proj_m = (I - qq')m
+            temp1 = q.T.conj() @ m
+            temp2 = q @ temp1
+            proj_m = m - temp2
```

Comment:
The changes here are comparable to those I made in ``nice_nullspace``. The runtime has improved to $O(MN^2)$ from the original $O(MN^3)$.